app/vmselect/netstorage/netstorage.go

```diff
@@ -1575,6 +1575,9 @@ func ProcessSearchQuery(qt *querytracer.Tracer, denyPartialResponse bool, sq *st
 					"to use more specific label filters in order to select lower number of series", *maxSamplesPerQuery),
 			}
 		}
+		if deadline.Exceeded() {
```

Comment:
I don't understand how this check improves deadline handling comparing to the current implementation, which works in the following way:

- The deadline is set at the connection level for both reading and writing to the connection before sending the request to `vmstorage` - see https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5b7f40907e4abd508b7ef61f4e94452902e3adbe/app/vmselect/netstorage/netstorage.go#L2096 . This connection level deadline exceeds the actual deadline by 2 seconds, so `vmstorage` could return `request timeout` error on itself and the connection to it could be re-used for new requests. Otherwise the connection to `vmstorage` should be closed, since it becomes broken after the read deadline is exceeded.

- The request is sent to `vmstorage`, the response data blocks are read in streaming mode and are passed to the `processBlock()` callback. See [`storageNode.processSearchQueryOnConn()` function](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5b7f40907e4abd508b7ef61f4e94452902e3adbe/app/vmselect/netstorage/netstorage.go#L2584-L2629). If the deadline is exceeded when reading the response blocks, then the `deadline exceeded` error is instantly returned from the `readBytes()` call at [this line](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5b7f40907e4abd508b7ef61f4e94452902e3adbe/app/vmselect/netstorage/netstorage.go#L2607). This instantly interrupts request processing.

In this case there is little sense in additional check for `deadline.Exceeded()` inside the `processBlock()` callback.

Reply:
The assumption here is that deadline is exceeded not because of network IO, but because of disk IO when dumping received data blocks on disk. The issue with exceeding the deadline was spotted on heavy-loaded vmselect, running many concurrent requests.